internal/tasks/publishertask.go

```diff
@@ -0,0 +1,152 @@
+package tasks
+
+import (
+	"context"
+
+	"github.com/razorpay/metro/internal/common"
+	"github.com/razorpay/metro/internal/nodebinding"
+	"github.com/razorpay/metro/internal/topic"
+	"github.com/razorpay/metro/pkg/logger"
+	"github.com/razorpay/metro/pkg/registry"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// PublisherTask implements the Watcher and maintains a pre-warmup.
+type PublisherTask struct {
+	id              string
+	registry        registry.IRegistry
+	topicCore       topic.ICore
+	nodeBindingCore nodebinding.ICore
+	topicWatchData  chan *struct{}
+}
+
+var topicCacheData map[string]bool = make(map[string]bool)
+var sem = make(chan int, 1)
+
+// NewPublisherTask creates PublisherTask instance
+func NewPublisherTask(
+	id string,
+	registry registry.IRegistry,
+	topicCore topic.ICore,
+	nodeBindingCore nodebinding.ICore,
+	options ...Option,
+) (ITask, error) {
+	publisherTask := &PublisherTask{
+		id:              id,
+		registry:        registry,
+		topicCore:       topicCore,
+		nodeBindingCore: nodeBindingCore,
+		topicWatchData:  make(chan *struct{}),
+	}
+
+	for _, option := range options {
+		option(publisherTask)
+	}
+
+	return publisherTask, nil
+}
+
+// Run the task
+func (pu *PublisherTask) Run(ctx context.Context) error {
+	var err error
+	var topicWatcher registry.IWatcher
+	err = pu.refreshCache(ctx)
+	if err != nil {
+		return err
+	}
+
+	twh := registry.WatchConfig{
+		WatchType: "keyprefix",
+		WatchPath: common.GetBasePrefix() + topic.Prefix,
+		Handler: func(ctx context.Context, pairs []registry.Pair) {
+			logger.Ctx(ctx).Infow("PublisherTask: topic watch handler data", "pairs", pairs)
+			pu.topicWatchData <- &struct{}{}
+		},
+	}
+
+	topicWatcher, err = pu.registry.Watch(ctx, &twh)
+	if err != nil {
+		return err
+	}
+
+	leadgrp, gctx := errgroup.WithContext(ctx)
+
+	// watch the Topic path for new topics and rebalance
+	leadgrp.Go(func() error {
+		watchErr := topicWatcher.StartWatch()
+		close(pu.topicWatchData)
+		return watchErr
+	})
+
+	// handle topic updates
+	leadgrp.Go(func() error {
+		for {
+			select {
+			case <-gctx.Done():
+				return gctx.Err()
+			case val := <-pu.topicWatchData:
+				if val == nil {
+					continue
+				}
+				terr := pu.refreshCache(ctx)
```

Comment:
at a time, only one process will be executing refreshCache, do we really need to introduce wait in the underlying func..?